### ISSUE TITLE
Add retry wrapper

### ISSUE THREAD

**Issue by georgysavva** [jackc/pgx#701]
As I saw in older issues, you don't want to implement a retry framework inside the library, because it's hard to implement it right. Because you don't know when you can retry a network error and when not. And I totally agree with you, because consistency of the data is a very important thing.
But it can be implement at the application level, because it actually knows the nature of the query: it it a read only query or it's an idempotent write and it's safe to retry it. 
Many applications would benefit from the retry logic, because it adds a lot resilience to your application.

Why not to introduce some kind of a wrapper around the library, where you set all your retry settings during the initialization. And after, when you call any query you just pass a flag whether the query is retryable or not. if it's false the wrapper just call the underlying library method, if it's true it applies the retry logic on some set or errors.

**Comment by jackc** [comment]
The basic functionality would be easy to implement. But even at the application layer there are subtle decisions involved in retry logic.

I have no objection to something like what you are describing, but I suspect getting it rock solid would take more effort than it may at first appear. Given the significant risk of subtle bugs, I would want some good reports from production usage before merging into an official release.

At the moment I do not have the free time to allocate to this, but if someone else wants to implement it I could help review it.

**Comment by georgysavva** [comment]
It good to hear that you like the idea, I will try to implement the wrapper and test it in production when I have time.
Can you share some of those "subtle decisions"  that one should be aware of.

**Comment by jackc** [comment]
So I wasn't planning on working on this, but a few ideas came to me this morning and I poked at the idea for a little while. I didn't get anything working but figured it would be worth recording the directions I explored and the problems I ran into.

I thought I had an approach that would even be easy to integrate into the existing pool. The first step would be for the pool to automatically retry queries that return an error marked as safe to retry.

```go
func (p *Pool) Exec(ctx context.Context, sql string, arguments ...interface{}) (pgconn.CommandTag, error) {
	fn := func() (pgconn.CommandTag, error) {
		c, err := p.Acquire(ctx)
		if err != nil {
			return nil, err
		}
		defer c.Release()

		return c.Exec(ctx, sql, arguments...)
	}

	ct, err := fn()
	attempts := int32(1)
	for err != nil && pgconn.SafeToRetry(err) && attempts < p.p.Stat().MaxResources() {
		ct, err = fn()
		attempts += 1
	}

	return ct, err
}
```

`attempts < p.p.Stat().MaxResources()` provides a limit so it can't get stuck retrying forever.

If this approach were sound it could easily be expanded to have the loop condition check for a value set in the context allowing it to be retried. Something like:

`(pgconn.SafeToRetry(err) || ctx.Value("retry") != nil)`

> Can you share some of those "subtle decisions" that one should be aware of.

These are some of the issues I ran into or would have needed to consider.

1. How many retries are allowed? I bounded it to the size of the pool with`attempts < p.p.Stat().MaxResources()` condition. But it might be more reasonable to only retry once.
2. As you mentioned in #250, it might be nice to have a default timeout for queries. The natural place for that would be at the connection level. But if there is retry logic in the mix somewhere then these timeouts wouldn't work properly. e.g. A 5 sec default timeout is set, when run through the retry system it is tried 10 times -- total timeout is 50 sec -- presumably not what the user intended.
3. `Query` returns a `Rows`. I don't know of any way to retry once you've begun reading.
4. `QueryRow` doesn't return an error. The error is deferred until the result is `Scan`ed. To retry this would mean a different implementation of the `pgx.Row` interface that internalized all this.
5. In the case of a network error it is likely that all connections are broken. How much of a point is there to retry?
6. Should retries be attempted even if it means establishing a new connection? If the server is just gone then each attempt at a new connection could have a very long wait. And again, in case of network error it is likely that everything will fail. I would only expect recovery to be possible in a HA environment where new connections would go to a different server (but HA might need more than simple retry logic).
7. When a retry occurs and fails which error is returned to the calling application? The original error may be different than the retry error.
8. What about transactions? Many application processes contain multiple queries that must be run in a transaction.

...and several of these issues are intertwined such as 1, 5, and 6.

Simulating network errors to test retry logic is difficult and retry logic is rarely executed in production. So it is difficult to have confidence in a particular solution. I generally prefer fail-fast when experiencing a network error rather than trying to retry. But if it was absolutely needed, I would probably approach it from a higher layer. Maybe something like:

```go
func WithTransactionalRetry(ctx context.Context, f func(ctx context.Context, tx pgx.Tx) error) error {
// f is run in transaction and rerun if necessary
}
``` 

Anything interesting at the application level will often need to be run in a transaction. So this is superior to retrying at the individual query level. This also would handle issues 3, 4, and 8. `WithTransactionalRetry` could also take other arguments that control how many times to retry and when to give up.

**Comment by georgysavva** [comment]
Thanks a lot for such a long list of your considerations. I've given it some thoughts, and will answer each of them correspondently:
1. It should be configurable. And actually, there are a lot of other parameters that configure retries. I would take a third-party library that implement an exponential backoff retries: [this one](https://github.com/cenkalti/backoff) for example. And pass all configurable settings into it.
2. This default timeout set at the pool level, so yes, it's applicable to the single operation on the pool (or transaction). The wrapper don't call the pool directly, instead user pass a closure and manually call pool inside the closure (more details about that in 3, 4, 8). So the user shouldn't expect that this timeout is also applicable to the retries. He can restrict the number of retries (this will roughly restrict the total time spent for retries), or he can restrict the total retries time explicitly, all of this settings happen at the wrapper level during initialization. 
3. Reading from rows also involves network operations that can fail. Wrapper should operate on a higher level, that cover the reading from rows too. We can pass a closure to the wrapper (as you suggested in 8) and interact with the pool and rows inside the wrapper, if something fails wrapper checks the nature of the error and retry the whole closure:
```golang
pool, err := pgxpool.Connect(context.Background(), os.Getenv("DATABASE_URL"))
pgxRetrier := NewPgxRetrier(retrySettings ... )
...
ctx := context.TODO()
var results []map[string]interface{}
pgxRetrier.Do(ctx, func() error {
       rows, err := pool.Query(ctx, sql, args ... )
       if err!=nil { return err }
       defer rows.Close()
       results=results[:0]
       // Iterate through the result set
       for rows.Next() {
             var result map[string]interface{}
             err = rows.Scan(&result)
             if err != nil {
                 return err
             }          
       }
       // Any errors encountered by rows.Next or rows.Scan will be returned here
       if rows.Err() != nil {
            return err
       }
       return nil
})
```
So it does both, execute the query and scan the results and if something fails it retries everything. Of course it only possible if you want to scan everything at once. If you have a large result that you want to iterate over step by step, then you shouldn't include scanning in the closure and wrapper will retry only if the execution failed.
4. The same with `QueryRow`, you execute and scan inside the closure and report any error that happened during the scan to the wrapper.
5 and 6. The purpose of those retries is to be resilient to transient network errors and particular database server currently unavailable. And yes, it only makes sense in an HA environment, I have multiple database servers behind a load balancer and retried query should go to a different server. So when retrying it should acquire a new connection from the pool. If pool is empty it might establish a new connection, but this logic is policy of the pool and depends on the current resources of the pull. Actually, I am going to use pgx with [CockroachDB](https://www.cockroachlabs.com/) It's a highly available, distributed SQL database that fully compatible with Postgres protocol. I was advised to use your library [here](https://github.com/cockroachdb/cockroach-go/issues/59)
7. The retry library that I mentioned in 1 returns the last error. I would go with that, don't see any problems with that approach.
8. Yes, we should retry transaction as a whole, using another connection from the pool:
```golang
pool, err := pgxpool.Connect(context.Background(), os.Getenv("DATABASE_URL"))
pgxRetrier := NewPgxRetrier(retrySettings ... )
...
ctx := context.TODO()
pgxRetrier.Do(ctx, func() error {
  tx, err := pool.Begin(ctx)
  if err != nil {
      return err
  }
  defer tx.Rollback(ctx)

  _, err = tx.Exec(ctx, "insert into foo(id) values (1)")
  if err != nil {
      return err
  }

  err = tx.Commit(ctx)
  if err != nil {
      return err
  }
  return nil
})
```
Or we could introduce transaction-aware method (as you suggested in 8), and let it manage transaction begin, commit, rollback. But in this case we will need to pass the pool object to the method or to the retrier during the initialization:
```golang
pool, err := pgxpool.Connect(context.Background(), os.Getenv("DATABASE_URL"))
pgxRetrier := NewPgxRetrier(retrySettings ... )
...
ctx := context.TODO()
pgxRetrier.DoTransaction(ctx, pool, func(tx pgx.TX) error {
  _, err = tx.Exec(ctx, "insert into foo(id) values (1)")
  if err != nil {
      return err
  }

  return nil
})
```

Please let me know what do you think about this and describe all your concerns. It will help to develop the most general and stable approach.

**Comment by jackc** [comment]
Your last example is what looks best to me. For the sake of application correctness I think that using a transaction should be the general rule. But it'd be easy to also have a non-tx version if that proved necessary.

My only further suggestion is that the `PgxRetrier` type may not be needed. `DoTransaction` could be a standalone function that directly takes `retrySettings` as an argument. That would reduce the entire retry system interface to a single function.

**Comment by georgysavva** [comment]
The only reason for type pgxRetrier is to store retrySettings that a user can define only once. I think we can have both approaches live together. To give more convenience to users. 
Thanks a lot for your participation. I am glad that we both have the same vision. I will start working on it as a separate package inside pgx.

**Comment by georgysavva** [comment]
Btw, if you say that using a transaction should be the general rule. May be worth adding .RunTransaction() on the library level, regardless of the retries. To hide .Begin(), .Commit(), .Rollback() from the user:
```golang
pool, err := pgxpool.Connect(context.Background(), os.Getenv("DATABASE_URL"))
...
ctx := context.TODO()
pool.RunTransaction(ctx, func(tx pgx.TX) error {
  _, err = tx.Exec(ctx, "insert into foo(id) values (1)")
  if err != nil {
      return err
  }

  return nil
})
```

**Comment by bfontaine** [comment]
Almost three years later, is this issue still relevant?

**Comment by mitar** [comment]
Yes it is. :-) I also came here to see if there is already some code to automatically retry serializable transactions or if I should write something myself.

**Comment by jackc** [comment]
As mentioned above, it's very unlikely that something like this will land in core pgx. And definitely not without some positive reports from people using it in production.

However, for people who do need this functionality, it may be worth looking at https://failsafe-go.dev/retry/ at least for inspiration.

**Comment by mitar** [comment]
Thanks for the link, this looks great.

In my case, I ended up writing a simple wrapper myself. I understand that there is hard to provide something general for everyone, but many queries in my case (where I use serializable isolation level) might have some benefit of a simple retry mechanism because there is nothing which changes between retries. In fact, I am a bit confusing why there has to be a client-server communication here and delays because of this. So currently I do:

* start transaction (sends a message to the server)
* do the query (another message to the server)
* try to commit the query (another message to the server)

If serialization fails, the client has to repeat again those three messages.

So it might be more reasonable to move this retry logic into a [stored procedure](https://www.postgresql.org/docs/current/sql-createprocedure.html) which now exists in PostgreSQL and can work with transactions so that it tries to retry a few times automatically on the server, without having to have any client-server communication. I am surprised that PostgreSQL does not provide a mechanism for this already, but probably it is for the same reason as here - it is hard to make it general. But I still think that for most cases it is useful. (In my case I use stored functions which do a lot of logic, so my non-trivial queries are mostly `SELECT myFunction()`.

**Comment by advdv** [comment]
We migrated from from Cockroach to Postgres so we needed a replacement for the "github.com/cockroachdb/cockroach-go/v2/crdb/crdbpgxv5" package. 

I came up with the following, it might be usefull to others. @jackc if you have time, let me know if you see any obvious problems to this approach so others can use it (or not) with more comfort. We will use it in production but it's a low traffic SaaS so i'm not sure this fact will help.

Go
```
// Package pgxretry allows for code to run with pgx transactions while retrying on serialization errors.
package pgxretry

import (
	"context"
	"errors"
	"fmt"

	"github.com/failsafe-go/failsafe-go"
	"github.com/failsafe-go/failsafe-go/retrypolicy"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgconn"
)

// number of retry attempts when a serializable error has occurred.
const maxSerializationErrorRetries = 50

// Conn abstracts pgx transactions creators: pgx.Conn and pgxpool.Pool.
type Conn interface {
	Begin(ctx context.Context) (pgx.Tx, error)
	BeginTx(ctx context.Context, txOptions pgx.TxOptions) (pgx.Tx, error)
}

// ExecuteTx runs fn inside a transaction and retries it as needed. On
// non-retryable failures, the transaction is aborted and rolled back; on
// success, the transaction is committed.
//
// conn can be a pgx.Conn or a pgxpool.Pool.
func ExecuteTx(
	ctx context.Context, conn Conn, txOptions pgx.TxOptions, fnc func(pgx.Tx) error,
) error {
	retry := retrypolicy.Builder[any]().HandleIf(func(_ any, err error) bool {
		var pgErr *pgconn.PgError
		if !errors.As(err, &pgErr) {
			return false
		}

		// https://www.postgresql.org/docs/current/mvcc-serialization-failure-handling.html
		return pgErr.Code == "40001"
	}).WithMaxRetries(maxSerializationErrorRetries).Build()

	//nolint: wrapcheck
	return failsafe.NewExecutor(retry).
		WithContext(ctx).
		RunWithExecution(func(exec failsafe.Execution[any]) (err error) { //nolint:contextcheck
			ctx := exec.Context()
			tx, err := conn.BeginTx(ctx, txOptions)
			if err != nil {
				return fmt.Errorf("failed to begin tx: %w", err)
			}

			defer tx.Rollback(ctx)

			if err := fnc(tx); err != nil {
				return err
			}

			if err := tx.Commit(ctx); err != nil {
				return fmt.Errorf("failed to commit: %w", err)
			}

			return err
		})
}
```

**Comment by jackc** [comment]
@advdv I don't see any obvious problems. Just be aware that the function you pass in doesn't do anything other than database operations as those might not be safe to retry.
